src/dinit-iostream.cc

```diff
@@ -0,0 +1,724 @@
+#include <limits>
+#include <cstdio>
+#include <cstring>
+
+#include <sys/types.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include <dinit-iostream.h>
+#include <baseproc-sys.h>
+
+// See dinit-iostream.h for inforamtion about the usage
+
+namespace {
+
+// A simple trick for calculating how much digits can be in a type with compile-time integer
+// logarithm like function.
+template <typename T> constexpr unsigned type_max_numdigits(T num = std::numeric_limits<T>::max(),
+        unsigned pow = 0)
+{
+    return (num == 0) ? pow : type_max_numdigits(num / 10, pow + 1);
+}
+
+// We should have space for sign in signed variables.
+constexpr unsigned UINT_MAX_CHARS = type_max_numdigits<unsigned>();
+constexpr unsigned INT_MAX_CHARS = type_max_numdigits<int>() + 1;
+constexpr unsigned ULONG_MAX_CHARS = type_max_numdigits<unsigned long>();
+constexpr unsigned LONG_MAX_CHARS = type_max_numdigits<long>() + 1;
+
+} /* anonymous namespace */
+
+namespace dio {
+
+// class io_base
+
+streambuf *io_base::get_buf() noexcept
+{
+    return buf.get();
+}
+
+bool io_base::is_open() noexcept
+{
+    return (fd >= 0);
+}
+
+void io_base::set_fd(const int newfd) noexcept
+{
+    fd = newfd;
+}
+
+int io_base::get_fd() noexcept
+{
+    return fd;
+}
+
+// class ostream
+
+ssize_t ostream::put(const char *msg, size_t count) noexcept
+{
+    if (!good()) {
+        return -1;
+    }
+
+    // If the current buffer space is not enough to store the message, we fill the buffer as much
+    // as possible and subtract the number of characters used to fill the buffer from the "count"
+    // and add it to this variable.
+    size_t pre_buffered_chars = 0;
+
+    if (count == 0) {
+        return 0; // Null/Empty message
+    }
+    else if (count > buf->get_size()) {
+        ssize_t r = flush_nx();
+        if (r < 0) {
+            // io_error was set by flush_nx() call
+            return -1;
+        }
+        else while (count > 0) {
+            ssize_t res = bp_sys::write(fd, msg, count);
```

Comment:
Now it always copies all of the message via the buffer before writing it.

The point was to not try to write(...) the remaining portion of the message _if_ it could fit in the buffer. If it can't fit in the buffer, writing it directly is the right thing to do.

Reply:
Is it the right fix: https://github.com/davmac314/dinit/pull/263/commits/20c831dc1f8383057690e5a2bd43257e7d4891d0 ?

Reply:
It's better, but still not really optimal in at least one case (`put` large amount of data when buffer is empty). As I said, the whole point is to call `write` as few times as possible (technically that isn't perfectly optimal in all cases, but it's a good-enough heuristic).

So: think about a range of possible scenarios and whether the implementation has the fewest possible calls to `write`.